### ISSUE TITLE
Control flow analysis for discriminated unions breaks when identified by type guards

### ISSUE THREAD

**Issue by third774** [microsoft/TypeScript#33903]
**TypeScript Version:**  3.7-Beta

**Search Terms:**
type guard discriminated union

**Code**

```ts
const isSomeDiscriminatedUnion = (value: unknown): value is { type: "bar", lol: string } | { type: "baz", omg: string } => true

function test(value: unknown): string {
  if (isSomeDiscriminatedUnion(value)) {
    if (value.type === "bar") {
      // ERROR! type narrowing didn't happen. value is still of type
      // { type: "bar", lol: string } | { type: "baz", omg: string }
      return value.lol
    }
    // ERROR! type narrowing also didn't happen here
    return value.omg
  }
  return ""
}
```

**Expected behavior:**
Discriminated unions should be able to have their types narrowed after being identified by a type guard.

**Actual behavior:**
Discriminated unions do not have their types narrowed after being identified by a type guard.

**Playground Link:** 

https://www.typescriptlang.org/play/index.html?ts=3.7-Beta#code/MYewdgzgLgBAlhAyiAtgUwCIOAJziuMAQyjQBMBVMOcGAXhgAoA3IgGwFc0AuGDsANZgQAdzABKXq05p4EGAG8YUAJ4AHHjABEAIyI4tAGhhsQbXtDxgA5jAC+MAD6Ll6zbqIAvIzFTWLUFa2DnQAfMo4XACwAFCxAGb8wFA0YMpo0CzsXLz8QqISAUGKsTDw8UwIyOhYELj4hCTkVKlZMuLiJTFlZXAVbVwAdKoa9HQMHgadCqU9ZQD08zAAKm4wxDg4ooS2EAAWIBxsZDB7RGoaYIMw0lww+4fHMDqyIBUjaLNziy4fvJM+Uzme6BHb2L49HBoKAcHBpW5oQZAiF2CE-VajDZbERgh5HE5nC5oK43bKyPFPF6+d5uNFLJR-bR6bzGPxFMGo7qQ6Gw+FkwZ+WacspQmFw7RaWKcoA

**Related Issues:**

**Comment by jack-williams** [comment]
I have a fix for this here: #30593.

The issue tracking it is this: #30557.

**Comment by third774** [comment]
Oh, awesome! Thanks! 😄
